src/main/java/net/ilexiconn/llibrary/client/model/tabula/animation/AnimationPlayerAnimator.java

```diff
@@ -0,0 +1,95 @@
+import net.ilexiconn.llibrary.server.animation.IAnimatedEntity;
+import net.ilexiconn.llibrary.server.animation.NamedAnimation;
+import net.minecraft.entity.Entity;
+import net.minecraft.util.math.MathHelper;
+
+/**
+ * Tabula Model animator that can play Tabula animations. USE {@link NamedAnimation} !!
+ * @author jglrxavpok
+ * @param <T> the entity type, needs to be an {@link IAnimatedEntity}
+ */
+public class AnimationPlayerAnimator<T extends Entity&IAnimatedEntity> implements ITabulaModelAnimator<T> {
+    @Override
+    public void setRotationAngles(TabulaModel model, T entity, float limbSwing, float limbSwingAmount, float ageInTicks, float rotationYaw, float rotationPitch, float scale) {
+        Animation anim = entity.getAnimation();
+        TabulaAnimationContainer animation;
+        if(anim instanceof NamedAnimation) {
+            animation = model.getAnimation(((NamedAnimation) anim).getName());
+        } else {
+            return;
+        }
+
+        if(animation == null)
+            return;
+
+        model.resetToDefaultPose();
+
+        preAnimationCallback(model, entity, limbSwing, limbSwingAmount, ageInTicks, rotationYaw, rotationPitch);
+
+        animation.getComponents().entrySet().forEach(entry -> {
+            for(TabulaAnimationComponentContainer component : entry.getValue()) {
+                applyComponent(component, model.getCubeByIdentifier(entry.getKey()), entity);
+            }
+        });
+
+        postAnimationCallback(model, entity, limbSwing, limbSwingAmount, ageInTicks, rotationYaw, rotationPitch);
+    }
+
+    /**
+     * Callback to use if you want to change model parts **before animating** but **after resetting to the default pose**. Uses the same argument as {@link #setRotationAngles(TabulaModel, Entity, float, float, float, float, float, float)}
+     */
+    protected void preAnimationCallback(TabulaModel model, T entity, float limbSwing, float limbSwingAmount, float ageInTicks, float rotationYaw, float rotationPitch) { }
+
+    /**
+     * Callback to use if you want to change model parts **after animating**. Uses the same argument as {@link #setRotationAngles(TabulaModel, Entity, float, float, float, float, float, float)}
+     */
+    protected void postAnimationCallback(TabulaModel model, T entity, float limbSwing, float limbSwingAmount, float ageInTicks, float rotationYaw, float rotationPitch) { }
+
+    /**
+     * Animates a given cube with the component
+     * @param component the animation component
+     * @param cube the cube to animate
+     * @param entity the animated entity
+     */
+    private void applyComponent(TabulaAnimationComponentContainer component, AdvancedModelRenderer cube, T entity) {
+        int tick = entity.getAnimationTick();
+        double progress = (tick - component.getStartKey()) / (double)component.getLength();
+        progress = MathHelper.clamp(progress, 0.0, 1.0);
+        if(tick >= component.getStartKey()) {
```

Comment:
Is `tick` not always `>= startKey`?

Reply:
Ah right. But does this not cause the animation to be applied twice? It's being applied here and then also afterwards scaled with `progress`.

Reply:
`progress` is used for the 'change' parameters while `tick >= startkey` use the 'offset' parameters

Reply:
Ahhh. Didn't notice that.